### ISSUE TITLE
How do you auto reload pdf after compiling?

### ISSUE THREAD

**Issue by xiebruce** [lhl2617/VSLilyPond#355]
I was able to compile the `test.ly` by pressing `command+s` on my Mac and it generate a `test.pdf` file, I open this pdf file in a split window on vscode. 
![image](https://user-images.githubusercontent.com/13575500/144744656-b7ba0285-7449-4d8c-b94e-ccb018152e7e.png)

As you known, we need to repeat this procedure: write code→save(compile)→preview, but I can't do this, every time I save(compile) the `test.ly` file, the `test.pdf` file not auto reloading, I have to close it and reopen to see the change.

How can I solve this problem?

vslilypond
```
v1.7.3
```

vscode
```
Version: 1.62.3 (Universal)
Commit: ccbaa2d27e38e5afa3e5c21c1c7bef4657064247
Date: 2021-11-17T07:59:13.865Z (2 wks ago)
Electron: 13.5.2
Chrome: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Darwin x64 20.6.0
```

macOS
```
Big Sur 11.5.1 (20G80)
```

**Comment by Apostolique** [comment]
Did you get the [LilyPond PDF Preview](https://marketplace.visualstudio.com/items?itemName=lhl2617.lilypond-pdf-preview) extension for vscode?

**Comment by xiebruce** [comment]
> Did you get the [LilyPond PDF Preview](https://marketplace.visualstudio.com/items?itemName=lhl2617.lilypond-pdf-preview) extension for vscode?

No, I didn't, because I think it's included in VSLilyPond, see [here](https://music.stackexchange.com/a/114500/83332) and [here](https://marketplace.visualstudio.com/items?itemName=lhl2617.lilypond-pdf-preview)

![image](https://user-images.githubusercontent.com/13575500/144755057-836357ac-1db6-4d76-b802-5bfddc119a0d.png)

![image](https://user-images.githubusercontent.com/13575500/144755023-00f3463b-bb35-4fc1-8187-1dbc30e2407c.png)

**Comment by Apostolique** [comment]
But were you using it?

**Comment by xiebruce** [comment]
> But were you using it?

I think yes
![image](https://user-images.githubusercontent.com/13575500/144756937-974d057d-d62e-485c-98ee-aac4ccf25241.png)

And I found "LilyPond PDF Preview" has been installed automatically
![image](https://user-images.githubusercontent.com/13575500/144756946-9893e0e2-f818-48a0-9d5a-238475647d2e.png)

when I double click on the pdf file that generated by lilypond in vscode, it's opened by the LilyPond PDF Preview plugin
![reopen by pdf lilypond previewer](https://user-images.githubusercontent.com/13575500/144757254-db3e859b-ebff-47c6-9ea9-087d5ccf2583.gif)

**Comment by lhl2617** [comment]
This is strange, on my Mac with 11.6 Big Sur, VSLilyPond 1.60  I can’t reproduce this issue

**Comment by xiebruce** [comment]
> This is strange, on my Mac with 11.6 Big Sur, VSLilyPond 1.60 I can’t reproduce this issue

Can you try the latest version of vsccode(`Version: 1.62.3 (Universal)`) and VSLilyPond(`v1.7.3`)? maybe you can reproduce this issue.

Here is my demo, if it is as expected, the pdf will auto reload after I save(compile) the `test.ly` right? But it didn't!
![vslilypond-pdf-not-auto-reloading](https://user-images.githubusercontent.com/13575500/144797007-a7129267-1fcc-496b-90e3-5c5f60576966.gif)

**Comment by Apostolique** [comment]
Do the other features work like clicking notes to go from pdf to code and from code to pdf?

**Comment by xiebruce** [comment]
> Do the other features work like clicking notes to go from pdf to code and from code to pdf?

Pdf to code: Yes, it works
Code to pdf: No, do I need to hold any key before clicking? Actually I tried hold option, cmd, shift, none of these works.

![click-pdf-to-code_optimized](https://user-images.githubusercontent.com/13575500/144818245-9f35f785-2c97-4a2d-b64d-5185d270f621.gif)

**Comment by Apostolique** [comment]
I forgot that was something from the command pallet (Ctrl + Shift + P): `LilyPond PDF Preview: Go to PDF location from Cursor`.

If I open the vscode Output panel and select `VSLilyPond: PDF Preview`, when I save a LilyPond file I see this log:

```
[LOG]: [WEBVIEW]: reload
[LOG]: [WEBVIEW]: pagesinit
[LOG]: [WEBVIEW]: allPagesLoaded
[LOG]: Clearing links for f:\Compositions\Sample01.pdf
[LOG]: Links cleared for f:\Compositions\Sample01.pdf
[LOG]: [WEBVIEW]: Sent clear-links
[LOG]: [WEBVIEW]: Finished handling textedits
[LOG]: [WEBVIEW]: Received link-register-ready
[LOG]: [WEBVIEW]: Finished registering links
```

**Comment by xiebruce** [comment]
> I forgot that was something from the command pallet (Ctrl + Shift + P): `LilyPond PDF Preview: Go to PDF location from Cursor`.
> 
> If I open the vscode Output panel and select `VSLilyPond: PDF Preview`, when I save a LilyPond file I see this log:
> 
> ```
> [LOG]: [WEBVIEW]: reload
> [LOG]: [WEBVIEW]: pagesinit
> [LOG]: [WEBVIEW]: allPagesLoaded
> [LOG]: Clearing links for f:\Compositions\Sample01.pdf
> [LOG]: Links cleared for f:\Compositions\Sample01.pdf
> [LOG]: [WEBVIEW]: Sent clear-links
> [LOG]: [WEBVIEW]: Finished handling textedits
> [LOG]: [WEBVIEW]: Received link-register-ready
> [LOG]: [WEBVIEW]: Finished registering links
> ```

There's no `LilyPond PDF Preview: Go to PDF location from Cursor` option on my vscode
![go-to-pdf_optimized](https://user-images.githubusercontent.com/13575500/144842198-ddd09a77-e9aa-4e48-8a26-ff8d1423eb63.gif)

I tried to uninstall and reinstall it, but I can't
![image](https://user-images.githubusercontent.com/13575500/144841756-bfc8a1a0-1efd-491d-9dd8-405f687c525b.png)

**Comment by Apostolique** [comment]
You got my instructions backward. It's not `VSLilyPond: Go to PDF...` It's `LilyPond PDF Preview: Go to PDF location from Cursor`.

Also look in your Output for VSLilyPond: PDF Preview: 
![image](https://user-images.githubusercontent.com/2766423/144843189-0ab1c49f-1a0c-4565-a671-229d3385b9a3.png)

**Comment by xiebruce** [comment]
> You got my instructions backward. It's not `VSLilyPond: Go to PDF...` It's `LilyPond PDF Preview: Go to PDF location from Cursor`.
> 
> Also look in your Output for VSLilyPond: PDF Preview: ![image](https://user-images.githubusercontent.com/2766423/144843189-0ab1c49f-1a0c-4565-a671-229d3385b9a3.png)

Sorry, I found it, but I can't see any log(including reload log) in the `VSLilyPond: PDF Preview` console
![no-reload-log](https://user-images.githubusercontent.com/13575500/144853586-4a6f4fc2-ba87-401b-b93e-b74634421b5a.gif)

And I've tried uninstall all lilypond relative extensions and reinstall vslilypond, still the same, no reloading operations can be found.

**Comment by xiebruce** [comment]
@Apostolique @lhl2617 Hi, according to the info I provided above, do you guys know what's happening on my vscode or vslilypond ?

**Comment by Apostolique** [comment]
I'm not sure but since it doesn't work you can find any PDF reader that has auto reload for now.

I use SumatraPDF but that's for Windows. I'm sure you can find a similar one for mac.

**Comment by xiebruce** [comment]
> I'm not sure but since it doesn't work you can find any PDF reader that has auto reload for now.
> 
> I use SumatraPDF but that's for Windows. I'm sure you can find a similar one for mac.

Yep, the build-in App "Preview.app" on Mac can do this, but I have to press `command+tab`(same as `alt+tab` on Win) to see the change.

**Comment by lhl2617** [comment]
I'm unable to repro your bug, can you try opening a dev console in VSCode (https://stackoverflow.com/questions/30765782/what-is-the-use-of-the-developer-tools-in-vs-code) and see if anything is wrong there?

**Comment by xiebruce** [comment]
@lhl2617 
```
Uncaught (in promise) ReferenceError: Cannot access 'd' before initialization
    at pdf.viewer.js:43
```

![image](https://user-images.githubusercontent.com/13575500/145836833-4c0109bc-7477-4c6f-85a1-72aac97d1ba1.png)

**Comment by Dr-ZeeD** [comment]
I seem to be having the same issue. It had worked for me on Big Sur before. After a completely clean install of VS Code and VSLilypond, the auto-refresh of the PDF preview window (which *really* uses the LilyPond PDF Preview that comes with VSLilyPond) does not work anymore.

**Comment by Dr-ZeeD** [comment]
Interestingly, I found a situation that results in the PDF within VSCode to refresh:

1. Open a lilypond file and its preview within VSCode. Also, open a Preview (the app) window of the PDF.
2. Modify the lilypond file and compile it. Right now, neither the VSCode preview nor the Preview should reflect the changes.
3. Click on the Preview window. It should refresh and show the changes. After a brief period, the VSCode preview will also refresh and show the changes.

**Comment by xiebruce** [comment]
> Interestingly, I found a situation that results in the PDF within VSCode to refresh:
> 
> 1. Open a lilypond file and its preview within VSCode. Also, open a Preview (the app) window of the PDF.
> 2. Modify the lilypond file and compile it. Right now, neither the VSCode preview nor the Preview should reflect the changes.
> 3. Click on the Preview window. It should refresh and show the changes. After a brief period, the VSCode preview will also refresh and show the changes.

I made a test in your condition, it turns out that the we should switch to another app then switch back to vscode after saving the code, only by doing this, the PDF inside vscode(opened by LilyPond PDF Preview) can reload. This has nothing to do with the macOS app "Preview.app".

**Comment by Dr-ZeeD** [comment]
I've just tried it with "not Preview.app". Does not work. So far I have only gotten it to work by clicking on Preview.app with the same file.

**Comment by Tobertet** [comment]
I have the same problem, and I am using Linux.

Whenever I save, I can see the logs in the Output panel and the PDF is being generated. However, the PDF Preview is not refreshing the output.

**Comment by jeromepochat** [comment]
I have same issue.

Code OSS > Help > About :

Version: 1.64.2
Commit: f80445acd5a3dadef24aa209168452a3d97cc326
Date: 2022-02-10T19:07:05.314Z
Electron: 13.6.9
Chromium: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.40-electron.0
OS: Linux x64 5.16.11-arch1-1

**Comment by snowskeleton** [comment]
I'm having the same problem.

macOS v12.4 Monterey
VSCode v1.69.1 (Universal)
VSLilyPond v1.7.3
LilyPond PDF Preview v0.2.8 (installed with VSLilyPond)

Switching to Preview.app and back updates the PDF both in Preview.app and in LilyPond PDF Viewer. Switching to an app other than Preview.app doesn't seem to trigger the reload. I also tried opening the PDF (after saving/compiling) with QuickLook (pressing Space bar on macOS in the Finder). QuickLook showed me the updated PDF, but neither Preview.app nor LilyPond PDF Viewer updated (until focusing Preview.app, of course).

**Comment by snowskeleton** [comment]
Update: I found that LilyPond PDF Viewer will automatically update if it's let sit long enough. In my experience, it's usually less than a minute, but not consistent. I can not consistently replicate this.

**Comment by clojj** [comment]
Hi @lhl2617 ..thanks for this fantastic plugin !

Locally debugging the PDF Preview Extension, I found that (on Linux) compiling with lilypond will generate a _new_ PDF file (with the same name).
Therefore when I created a listener for *creation* event like this:
```ts
        this._watcher.onDidCreate((e) => {
          if (e.toString() === this.resource.toString()) {
            this.reload()
          }
        })
```
here
https://github.com/lhl2617/VSLilyPond-PDF-preview/blob/master/src/extension/pdf-preview.ts#L62
..automatical reload works fine !

So this yields the somewhat weird question, if all OS's lilypond command treat PDF compilation the same or different (?)

If you want I can create a PR and try to cover _all_ events

**Comment by zzzeek** [comment]
hi folks, I have this bug.   I also don't have ready ability to hack on the plugin itself.   

so to work around this, I open my console, select "terminal", and on my Fedora linux machine (things may have to be different on OSX) I run this while loop to refresh the pdf each time it gets recreated, which seems to wake up the pdf viewer:

```bash
while [ 1 ]; do sleep 1; find . -name "*.pdf" -mmin -0.02  -exec sh -c "echo 'detected changes on {}'; touch {}; sleep 1 " \;   ; done;
```

That will run every second, and it will:  1. run a find for .pdf files that were modified less than .02 minutes ago 2. do a touch 3. the sleep another second per "touch", so that it doesn't keep touching the file over and over again.    

I organize that into a script `refresh.sh` and just have it wherever my .ly files are.

(noting I just tried lilypond for the first time about an hour ago)

**Comment by ghost** [comment]
> Hi @lhl2617 ..thanks for this fantastic plugin !
> 
> Locally debugging the PDF Preview Extension, I found that (on Linux) compiling with lilypond will generate a _new_ PDF file (with the same name). Therefore when I created a listener for _creation_ event like this:
> 
> ```ts
>         this._watcher.onDidCreate((e) => {
>           if (e.toString() === this.resource.toString()) {
>             this.reload()
>           }
>         })
> ```
> 
> here https://github.com/lhl2617/VSLilyPond-PDF-preview/blob/master/src/extension/pdf-preview.ts#L62 ..automatical reload works fine !
> 
> So this yields the somewhat weird question, if all OS's lilypond command treat PDF compilation the same or different (?)
> 
> If you want I can create a PR and try to cover _all_ events

Thank you for offering the solution, I did this on my Mac and it worked!
![image](https://github.com/lhl2617/VSLilyPond/assets/61086484/12d5f5b0-d4a4-47a3-8374-0fb1bf0f4021)

**Comment by Noelind** [comment]
First of all thank you @lhl2617 for making this possible in the first place! This is my first time venturing in Lilypond territory and I figured it'd be good fit for me given my familiarity with VSCode and TeX. And since I've set this up, I can't even think of going back to GUI based sheet editors.

I realize this is an old issue but I wanted to share a fix that worked for me regarding the auto-reload of the .pdf file. Note that I'm doing this on a MacOS Sonoma 14.5 machine. Shouldn't be too different though on other systems. @clojj and @leiyi-0612 comments were immensely helpful for fixing this 🙏. _(I'm pretty much repeating @leiyi-0612 instructions step by step)_

1. Open `pdf-preview.js` located at `/Users/{username}/.vscode/extensions/lhl2617.lilypond-pdf-preview-0.2.8/out/extension/pdf-preview.js`
2. Find statement `const watcher = this._register(vscode.workspace.createFileSystemWatcher(resource.fsPath));`
3. Paste these lines below that statement and save.
```javascript
        this._register(watcher.onDidCreate((e) => {
          if (e.toString() === this.resource.toString()) {
            this.reload()
          }
        }));
``` 
4. Reload VSCode and edit/save a .ly file.

------------------------

Extra thoughts: Not an expert on this but my guess as to why `onDidChange` isn't working is because the .pdf file is being built from scratch everytime a .ly file is saved. Meaning there's a brief period of time where the file doesn't exist and `onDidChange` can't locate it(??)

**Comment by jacopogrecodalceo** [comment]
Hello,
I’m still encountering the same issue: the PDF preview does not auto-refresh after LilyPond compilation.

The required configuration lines are already present, so I believe the problem lies elsewhere.

From the console, I see the following messages:

```
Warning: Setting up fake worker.
Could not load content for https://file+.vscode-resource.vscode-cdn.net/Users/o/.cursor/extensions/lhl2617.lilypond-pdf-preview-0.2.8/src/pdfjs-custom/main.ts (Connection error: net::ERR_NAME_NOT_RESOLVED)
```
It seems like the `src/pdfjs-custom/main.ts` component is failing to load properly.

Do you have any suggestions?
